lib/evmone/eof.cpp

```diff
@@ -90,37 +151,261 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+
+            const auto count = code[i + 1];
+
+            const auto post_offset = 1 + 1 /* count */ + count * REL_OFFSET_SIZE /* tbl */;
+
+            for (size_t k = 0; k < count * REL_OFFSET_SIZE; k += REL_OFFSET_SIZE)
+            {
+                const auto rel_offset = read_int16_be(&code[i + 1 + 1 + static_cast<uint16_t>(k)]);
+                const auto jumpdest = static_cast<int>(i + post_offset) + rel_offset;
+                if (jumpdest < 0 || static_cast<size_t>(jumpdest) >= code_size)
+                    return false;
+                rjumpdests.push_back(static_cast<size_t>(jumpdest));
+            }
+
+            const auto rjumpv_imm_size = size_t{1} + count * REL_OFFSET_SIZE;
+            std::fill_n(
+                immediate_map.begin() + static_cast<ptrdiff_t>(i) + 1, rjumpv_imm_size, true);
+            i += rjumpv_imm_size;
+            continue;
+        }
+
+        const auto imm_size = instr::traits[op].immediate_size;
+        std::fill_n(immediate_map.begin() + static_cast<ptrdiff_t>(i) + 1, imm_size, true);
+        i += imm_size;
+    }
+
+    // Check relative jump destinations against immediate locations.
+    for (const auto rjumpdest : rjumpdests)
+        if (immediate_map[rjumpdest])
+            return false;
+
+    return true;
+}
+
+std::pair<EOFValidationError, int32_t> validate_max_stack_height(
+    bytes_view code, size_t func_index, const std::vector<EOFCodeType>& funcs_in_outs)
+{
+    assert(!code.empty());
+
+    // Changing this type to smaller integer (uint16_t) have in mind that it can overflow.
+    std::vector<int32_t> stack_heights = std::vector<int32_t>(code.size(), -1);
+    std::stack<size_t> worklist;
+
+    int32_t stack_height = 0;
+    stack_heights[0] = funcs_in_outs[func_index].inputs;
+    worklist.push(0);
+
+    size_t i = 0;
+    Opcode opcode = {};
+    std::vector<size_t> successors;
+    while (!worklist.empty())
+    {
+        i = worklist.top();
+        opcode = static_cast<Opcode>(code[i]);
+        worklist.pop();
+
+        auto stack_height_required = instr::traits[opcode].stack_height_required;
+        auto stack_height_change = instr::traits[opcode].stack_height_change;
+
+        if (opcode == OP_CALLF)
+        {
+            auto fid = read_uint16_be(&code[i + 1]);
+
+            if (fid >= funcs_in_outs.size())
+                return {EOFValidationError::invalid_code_section_index, 0};
+
+            stack_height_required = static_cast<int8_t>(funcs_in_outs[fid].inputs);
+            auto d = funcs_in_outs[fid].outputs - stack_height_required;
+            stack_height_change = static_cast<int8_t>(d);
+        }
+
+        stack_height = stack_heights[i];
+        assert(stack_height != -1);
+
+        if (stack_height < stack_height_required)
+            return {EOFValidationError::stack_underflow, -1};
+
+        successors.clear();
+
+        // immediate_size for RJUMPV depends on the code. It's calculated below.
```

Comment:
This comment doesn't seem to be related? This won't match for `RJUMPV`

Reply:
This comment clarifies why we are skipping RJUMPV. I rephrased to make it clearer.

Reply:
But this function skips both RJUMPV and RJUMP?

Reply:
So `RJUMP` is skipped because `next` is not a successor for `RJUMP`. The `successor` is calculated below. For `RJUMPI` `next` is a successor so it's not skipped. `RJUMPV` is skipped because `immediate_size` does not make sense for it. I'm adding better comment covering `RJUMP`

Reply:
Okay finally getting closer to understanding it. I think this if/else tree we have here could be restructured and made more logical:
```
if (opcode == OP_RJUMP || opcode == OP_RJUMPI)
...
else if (opcode == OP_RJUMPV)
...
else if (!instr::traits[opcode].is_terminating)
...
else
```